File: packages/people/android/src/main/java/io/capkit/people/error/PeopleError.kt
package io.capkit.people.error

/**
 * Native error model for the People plugin (Android).
 *
 * Architectural rules:
 * - Must NOT reference Capacitor APIs.
 * - Must NOT reference JavaScript directly.
 * - Must be throwable from the Implementation (Impl) layer.
 * - Mapping to JS-facing error codes happens ONLY in the Plugin layer.
 */
sealed class PeopleError(
  message: String,
) : Throwable(message) {
  // -----------------------------------------------------------------------------
  // Specific Error Types
  // -----------------------------------------------------------------------------

  /**
   * Feature or capability is not available due to device or configuration limitations.
   * Maps to the 'UNAVAILABLE' error code in JavaScript.
   */
  class Unavailable(
    message: String,
  ) : PeopleError(message)

  /**
   * The user cancelled an interactive flow (e.g., contact picker).
   * Maps to the 'CANCELLED' error code in JavaScript.
   */
  class Cancelled(
    message: String,
  ) : PeopleError(message)

  /**
   * Required permission was denied or not granted by the user.
   * Maps to the 'PERMISSION_DENIED' error code in JavaScript.
   */
  class PermissionDenied(
    message: String,
  ) : PeopleError(message)

  /**
   * Plugin failed to initialize or perform a required native operation.
   * Maps to the 'INIT_FAILED' error code in JavaScript.
   */
  class InitFailed(
    message: String,
  ) : PeopleError(message)

  /**
   * Invalid or malformed input was provided by the caller.
   * Maps to the 'INVALID_INPUT' error code in JavaScript.
   */
  class InvalidInput(
    message: String,
  ) : PeopleError(message)

  /**
   * Invalid or unsupported input type was provided to the native implementation.
   * Maps to the 'UNKNOWN_TYPE' error code in JavaScript.
   */
  class UnknownType(
    message: String,
  ) : PeopleError(message)

  /**
   * The requested resource does not exist (e.g., contact or group not found).
   * Maps to the 'NOT_FOUND' error code in JavaScript.
   */
  class NotFound(
    message: String,
  ) : PeopleError(message)

  /**
   * The operation conflicts with the current state (e.g., read-only group).
   * Maps to the 'CONFLICT' error code in JavaScript.
   */
  class Conflict(
    message: String,
  ) : PeopleError(message)

  /**
   * The operation did not complete within the expected time.
   * Maps to the 'TIMEOUT' error code in JavaScript.
   */
  class Timeout(
    message: String,
  ) : PeopleError(message)
}

File: packages/people/ios/Sources/PeoplePlugin/error/PeopleError.swift
import Foundation

/**
 Native error model for the People plugin (iOS).

 This enum represents all error categories that can be
 produced by the native implementation layer.

 Architectural rules:
 - Must NOT reference Capacitor
 - Must NOT reference JavaScript
 - Must be throwable from the Impl layer
 - Mapping to JS-facing error codes happens ONLY in the Plugin layer
 */
enum PeopleError: Error {

    /// Feature or capability is not available on this device or configuration
    case unavailable(String)

    /// The user cancelled an interactive flow (e.g., contact picker)
    case cancelled(String)

    /// Required permission was denied or not granted
    case permissionDenied(String)

    /// Plugin failed to initialize or perform a required operation
    case initFailed(String)

    /// The input provided to the plugin method is invalid or malformed
    case invalidInput(String)

    /// Invalid or unsupported input was provided
    case unknownType(String)

    /// The requested resource does not exist (e.g., contact or group not found)
    case notFound(String)

    /// The operation conflicts with the current state (e.g., read-only group)
    case conflict(String)

    /// The operation did not complete within the expected time
    case timeout(String)

    // MARK: - Human-readable message

    /**
     Human-readable error message.

     This message is intended to be passed verbatim
     to JavaScript via `call.reject(message, code)`.
     */
    var message: String {
        switch self {
        case .unavailable(let message):
            return message
        case .cancelled(let message):
            return message
        case .permissionDenied(let message):
            return message
        case .initFailed(let message):
            return message
        case .invalidInput(let message):
            return message
        case .unknownType(let message):
            return message
        case .notFound(let message):
            return message
        case .conflict(let message):
            return message
        case .timeout(let message):
            return message
        }
    }

    /// Standardized error code string for JS rejection.
    var errorCode: String {
        switch self {
        case .unavailable: return "UNAVAILABLE"
        case .cancelled: return "CANCELLED"
        case .permissionDenied: return "PERMISSION_DENIED"
        case .initFailed: return "INIT_FAILED"
        case .invalidInput: return "INVALID_INPUT"
        case .unknownType: return "UNKNOWN_TYPE"
        case .notFound: return "NOT_FOUND"
        case .conflict: return "CONFLICT"
        case .timeout: return "TIMEOUT"
        }
    }
}
